src/main/java/com/conveyal/datatools/manager/jobs/AutoDeployFeedJob.java

```diff
@@ -0,0 +1,133 @@
+package com.conveyal.datatools.manager.jobs;
+
+import com.conveyal.datatools.common.status.MonitorableJob;
+import com.conveyal.datatools.manager.auth.Auth0UserProfile;
+import com.conveyal.datatools.manager.controllers.api.DeploymentController;
+import com.conveyal.datatools.manager.models.Deployment;
+import com.conveyal.datatools.manager.models.FeedVersion;
+import com.conveyal.datatools.manager.models.OtpServer;
+import com.conveyal.datatools.manager.models.Project;
+import com.conveyal.datatools.manager.persistence.Persistence;
+import com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Auto deploy new feed version to OTP server if {@link Project#autoDeploy} is enabled and other conditions are met
+ * (e.g., feed version has no critical errors, active deployment is not in progress, there are no other feed fetches in
+ * progress, etc.). This job must run after {@link ValidateFeedJob} as it has a dependency on the outcome of
+ * {@link FeedVersion#hasCriticalErrors}.
+ */
+public class AutoDeployFeedJob extends MonitorableJob {
+    public static final Logger LOG = LoggerFactory.getLogger(ValidateFeedJob.class);
+
+    /**
+     * Projects to be considered for auto deployment
+     */
+    private List<Project> projects = new ArrayList<>();
+
+    /**
+     * A set of projects which have been locked by a instance of {@link AutoDeployFeedJob} to prevent repeat
+     * deployments.
+     */
+    private final Set<String> lockedProjects = Sets.newConcurrentHashSet();
```

Comment:
This needs to be a static variable. As it is, this gets instantiated each time a new job gets created, so it'll only be visible to the current instance of this class instead of to all instances.

Reply:
Also, the `Sets.newConcurrentHashSet` creates a ConcurrentHashMap. Although this does allow thread-safe reading and writing, it is inappropriate for placing locks because it only gets the value from the most recently completed transaction. Therefore, this set must be declared in a way that its methods are synchronized so that each thread trying to access this variable waits for any other thread to complete a transaction with this variable before proceeding. I recommend the following change:

```suggestion
    private static final Map<String, Boolean> lockedProjects = Collections.synchronizedMap(new HashMap<>());
```

Also, here's [a decent article](https://howtodoinjava.com/java/collections/hashmap/synchronize-hashmap/) explaining this in more detail.

Reply:
Updated.